backend/src/main/java/ai/verta/modeldb/entities/LineageEntity.java

```diff
@@ -9,44 +14,115 @@
 import javax.persistence.Table;
 
 @Entity
-@Table(name = "lineage")
+@Table(name = "lineage_versioning")
 public class LineageEntity implements Serializable {
 
   public LineageEntity() {}
 
-  public LineageEntity(LineageEntry input, LineageEntry output) {
-    inputExternalId = input.getExternalId();
-    inputType = input.getTypeValue();
-    outputExternalId = output.getExternalId();
-    outputType = output.getTypeValue();
+  public LineageEntity(Long id, LineageEntry input, LineageEntry output)
+      throws InvalidProtocolBufferException {
+    inputType = input.getDescriptionCase().getNumber();
+    switch (input.getDescriptionCase()) {
+      case EXPERIMENT_RUN:
+        inputExperimentId = input.getExperimentRun();
+        break;
+      case BLOB:
+        VersioningLineageEntry blob = input.getBlob();
+        inputRepositoryId = blob.getRepositoryId();
+        inputCommitSha = blob.getCommitSha();
+        inputLocation = ModelDBUtils.getStringFromProtoObject(Location.newBuilder().addAllLocation(blob.getLocationList()));
+    }
+    outputType = output.getDescriptionCase().getNumber();
+    switch (output.getDescriptionCase()) {
+      case EXPERIMENT_RUN:
+        outputExperimentId = output.getExperimentRun();
+        break;
+      case BLOB:
+        VersioningLineageEntry blob = output.getBlob();
+        outputRepositoryId = blob.getRepositoryId();
+        outputCommitSha = blob.getCommitSha();
+        outputLocation = ModelDBUtils.getStringFromProtoObject(Location.newBuilder().addAllLocation(blob.getLocationList()));
+    }
+    this.id = id;
   }
 
   @Id
-  @Column(name = "input_external_id")
-  private String inputExternalId;
+  Long id;
+
+  @Id
+  @Column(name = "input_experiment_id")
```

Comment:
Why should we store all the IO information in a single table? Sounds like trouble when you want to do a join, since you'll read a ton of information you don't care about. Shouldn't we have a better separation like:

1. a table for outputs of lineage + lineage entry ID + lineage entry type `(lineage_id int, entry_id int, entry_type int)`
2. a table for inputs of lineage + lineage entry ID + lineage entry type `(lineage_id int, entry_id int, entry_type int)`
3. a table for experiment run lineage entries `(entry_id int, experiment_run_id string)`
4. a table for commit sha entries `(entry_id int, repo_id int, commit_sha string, location string)`

cc @ravishetye for thoughts.

Reply:
to a minimum we should split  to 3 tables , and table 3 and 4 can be combined

Reply:
fixed